### DISCUSSION TITLE
Create full copy of metadata inside another database with another engine

### DISCUSSION THREAD

**Discussion by RedHotUnicorn** [sqlalchemy/sqlalchemy#9606] [Usage Questions] [answered]
I need to create a structure of existing DB without data inside tables.

It means that source and target could be different:

`mysql` -> `pg` , `pg` -> `oracle` , `sqlite` -> `whatever`...

So I choose Sqlalchemy to extracting the Metadata.

```python
import sqlalchemy as db
from sqlalchemy import create_engine, Table, Column, Integer, Unicode, MetaData, String, Text, update, and_, select, func, types
from sqlalchemy_utils import database_exists, create_database


srcEngine = db.create_engine('postgresql://LOGIN:PASS@host:5432/db') 
meta_data = db.MetaData(schema='public')

db.MetaData.reflect(meta_data,bind=srcEngine, resolve_fks=True )
m_order = meta_data.sorted_tables

for t in meta_data.tables.values():
    t.schema = None
    for c in t.c.values():
        try:
            c.type = c.type.as_generic()
        except  NotImplementedError as e:
            c.type = "TEXT"
        c.server_default=""

destEngine = db.create_engine("mysql+pymysql://LOGIN:PASS@host:3306/db_test")

if not database_exists(destEngine.url):
    create_database(destEngine.url)

destEngine._metadata = MetaData()
destEngine._metadata.tables = meta_data.sorted_tables

destEngine._metadata.create_all(bind=destEngine, checkfirst=True,tables=m_order)
```

How to convert CREATE TABLE statements to specific engine?

**Comment by CaselIT** [comment]
Hi,

In general tables are considered immutable, so to generate generic types it's best to use this event https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.DDLEvents.column_reflect 
At the moment it's not mentioned in the `MetaData.reflect` documentation but probably it should. I'll add a reference to make it more discoverable

> ```py
> destEngine._metadata = MetaData()
> destEngine._metadata.tables = meta_data.sorted_tables
> 
> destEngine._metadata.create_all(bind=destEngine, checkfirst=True,tables=m_order)
> ```

This is not needed, since the engine has no `_metadata` attribute. Furthermore a MetaData object is not connected with an engine, so just can just use the original metadata passing the destination engine to create: `db.MetaData.create_all(destEngine)`

**Reply by RedHotUnicorn** [reply]
Thanks for reply, but I don't get is there some path to automatically replace unsupported types ?

**Reply by CaselIT** [reply]
This section of the docs speaks of it https://docs.sqlalchemy.org/en/20/core/reflection.html#metadata-reflection-dbagnostic-types

**Reply by RedHotUnicorn** [reply]
So maybe I lost something in logic but as I understand method `as_generic()` that used in `column_reflect` only convert specific type to more global type like: tinyint -> integer

but what if we have global type that not supported in another DB.
For example I have table with that DDL:
```sql
CREATE TABLE `test` (
  `idtest` int(11) NOT NULL,
  `blob` blob,
  `datetime` datetime DEFAULT NULL,
  `binary` binary(16) NOT NULL,
  `tiny` tinyint(1) DEFAULT NULL,
  `date` date DEFAULT NULL,
  PRIMARY KEY (`idtest`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

after reflecting I get the next code inside python:
```sql
CREATE TABLE test (
        idtest INTEGER NOT NULL,      
        blob BLOB, 
        datetime DATETIME, 
        "binary" BINARY(16) NOT NULL, 
        tiny INTEGER, 
        date DATE, 
        PRIMARY KEY (idtest)
)
```

But this code will not work inside postgres.
Because there is no `BLOB`,  `DATETIME`  and `BINARY` types inside pg.

Maybe I can send rule of converting in some method?
thank you for your reply!

**Reply by CaselIT** [reply]
As generic is called also in your initial code.
Calling it on these types yield:
- BLOB-> LargeBinary. ok on PG
- DATETIME -> DateTime. ok on PG
- BINARY(16) -> does nothing since it's already generic, meaning that isn't a custom type from mysql). This does not exist on PG so you may need to use a custom path for this type. @zzzeek maybe postgresql should render bytea here? 

> Maybe I can send rule of converting in some method?

sure you can use as_generic as a starting point and then do custom things for each particular type

**Reply by RedHotUnicorn** [reply]
Hmm. 
It seems that it's not ok types for pg.
I'm running the last (15.2) pg instance and the code of creating the table fails just in this types of data.
According to [docs](https://www.postgresql.org/docs/current/datatype.html) there are no BLOB, DateTime data types (with that names).

**Reply by CaselIT** [reply]
Blob.as_generic() returns a LargeBinary that can be used on PG. Same for DATETIME -> DateTime.

You will need to find a type for BINARY, since there is not one similar in PG, but I guess you could use LargeBinary also here

**Reply by RedHotUnicorn** [reply, marked as answer]
thanks. 
Code that helps reflect columns
```python
@event.listens_for(meta_data, 'column_reflect')
def genericize_datatypes(inspector, tablename, column_dict):
    
    column_dict["default"] = None
    try:
        column_dict["type"] = column_dict["type"].as_generic()
    except  NotImplementedError as e:
        column_dict["type"] =  db.types.TEXT()
    if isinstance(column_dict["type"] ,  db.types.BINARY) or str(column_dict["type"]) == "BINARY":
        column_dict["type"] = db.types.TEXT()
    if isinstance(column_dict["type"] ,  db.types.DATETIME) or str(column_dict["type"]) == "DATETIME":
        column_dict["type"] = db.types.DATE()
    if isinstance(column_dict["type"]  ,  db.types.BLOB) or str(column_dict["type"] ) == "BLOB":
        column_dict["type"]  = db.types.TEXT()
    if isinstance(column_dict["type"]  ,  db.types.Enum) or str(column_dict["type"] ) == "Enum":
        column_dict["type"]  = db.types.TEXT()

    if getattr(column_dict["type"], 'collation', None) is not None:
        column_dict["type"].collation = None
    # print(column_dict["type"])
    # print(column_dict)
```

**Reply by CaselIT** [reply]
Seems ok,
Note that this does not seem correct
```py
if isinstance(column_dict["type"] ,  db.types.DATETIME) or str(column_dict["type"]) == "DATETIME":
        column_dict["type"] = db.types.DATE()
```
